### ISSUE TITLE
Misleading error message when there is a generic trait impl for all types implementing other trait

### ISSUE THREAD

**Issue by VFLashM** [rust-lang/rust#32704]
See example.
Error message mentions trait 'Base', even thought only 'Marker' trait is required.

``` rust
trait Base {}
trait Marker {}
impl<T: Base> Marker for T {}
impl Marker for i32 {}

fn requires_marker<T: Marker>(value: &T) {}

fn main() {
    let i: i32 = 0;
    let u: u32 = 0;
    requires_marker(&i); // fine, i32 is unrelated to Base
    requires_marker(&u); // error: the trait `Base` is not implemented for the type `u32`
}
```

**Comment by Mark-Simulacrum** [comment]
I believe this is a duplicate of https://github.com/rust-lang/rust/issues/28894; the compiler tries to see through the Marker bound and goes too far instead of stating that `Marker` isn't implemented.
